zebra-state/src/sled_state.rs

```diff
@@ -0,0 +1,195 @@
+//! The primary implementation of the `zebra_state::Service` built upon sled
+
+use std::{collections::HashMap, convert::TryInto, future::Future, sync::Arc};
+
+use zebra_chain::serialization::{ZcashDeserialize, ZcashSerialize};
+use zebra_chain::{
+    block::{self, Block},
+    parameters::Network,
+};
+
+use crate::{BoxError, Config, HashOrHeight, QueuedBlock};
+
+pub struct SledState {
+    /// Queued blocks that arrived out of order, indexed by their parent block hash.
+    queued_by_prev_hash: HashMap<block::Hash, QueuedBlock>,
+
+    hash_by_height: sled::Tree,
+    height_by_hash: sled::Tree,
+    block_by_height: sled::Tree,
+    // tx_by_hash: sled::Tree,
+    // utxo_by_outpoint: sled::Tree,
+    // sprout_nullifiers: sled::Tree,
+    // sapling_nullifiers: sled::Tree,
+    // sprout_anchors: sled::Tree,
+    // sapling_anchors: sled::Tree,
+}
+
+impl SledState {
+    pub fn new(config: &Config, network: Network) -> Self {
+        let db = config.sled_config(network).open().unwrap();
+
+        Self {
+            queued_by_prev_hash: HashMap::new(),
+            hash_by_height: db.open_tree(b"hash_by_height").unwrap(),
+            height_by_hash: db.open_tree(b"height_by_hash").unwrap(),
+            block_by_height: db.open_tree(b"block_by_height").unwrap(),
+            // tx_by_hash: db.open_tree(b"tx_by_hash").unwrap(),
+            // utxo_by_outpoint: db.open_tree(b"utxo_by_outpoint").unwrap(),
+            // sprout_nullifiers: db.open_tree(b"sprout_nullifiers").unwrap(),
+            // sapling_nullifiers: db.open_tree(b"sapling_nullifiers").unwrap(),
+        }
+    }
+
+    /// Queue a finalized block to be committed to the state.
+    pub fn queue(&mut self, queued_block: QueuedBlock) {
+        let prev_hash = queued_block.block.header.previous_block_hash;
+        self.queued_by_prev_hash.insert(prev_hash, queued_block);
+    }
+
+    pub fn process_queue(&mut self) {
+        // Cloning means the closure doesn't hold a borrow of &self,
+        // conflicting with mutable access in the loop below.
+        let hash_by_height = self.hash_by_height.clone();
+        let tip_hash = || {
+            read_tip(&hash_by_height)
+                .expect("inability to look up tip is unrecoverable")
+                .map(|(_height, hash)| hash)
+                .unwrap_or(block::Hash([0; 32]))
+        };
+
+        while let Some(queued_block) = self.queued_by_prev_hash.remove(&tip_hash()) {
+            self.commit_finalized(queued_block)
+        }
+    }
+
+    /// Commit a finalized block to the state. It's the caller's responsibility
+    /// to ensure that blocks are committed in order.
```

Comment:
comment looks out of date

Reply:
This function does have to be called in-order, which is why it's not `pub` -- it's called as part of `process_queue`, which makes sure that it's called in order.  I can update the comments to clarify.

Reply:
Fixed in 8b2142929641970fbf2b8ec20be6b1c3108e1d07